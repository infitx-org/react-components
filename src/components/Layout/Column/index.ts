export { default } from "./Column";
export type { ColumnProps } from "./Column";
