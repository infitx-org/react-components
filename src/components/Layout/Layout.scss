@use "sass:map";
@use "../../scss/components/any.scss";
@use "../../scss/themes.scss";

.rc-layout {
  @include any.rc-component;

  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
}

.rc-layout__content {
  @include any.rc-component;

  flex: 1 1 auto;
  max-width: 100%;
  display: flex;
  flex-direction: row;
  min-width: 0;
  overflow: hidden;
}

.rc-layout__page {
  @include any.rc-component;

  overflow-x: hidden;
  flex: 1 1 auto;
  padding: 20px;
}

.rc-layout__side-menu {
  @include any.rc-component;

  flex: 0 0 200px;
  border-right: 1px solid #ccc;
  background: #eee;
}
