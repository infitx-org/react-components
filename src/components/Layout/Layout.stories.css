.row {
  position: relative;
  background-color: #ddd;
  border-radius: 5px;
  height: 100px;
  width: 100%;
  margin: 5px;
}

.column {
  position: relative;
  background-color: #ddd;
  border-radius: 5px;
  width: 100px;
  height: 400px;
  margin: 5px;
}

.title {
  position: absolute;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-family: Verdana;
  color: #06c;
  text-shadow: 0px 0px 4px #fff;
}

.block--small {
  background: #333;
  margin: 5px;
  height: 20px;
  width: 20px;
  border-radius: 10%;
}

.block--big {
  background: #666;
  margin: 5px;
  height: 40px;
  width: 40px;
  border-radius: 10%;
}
