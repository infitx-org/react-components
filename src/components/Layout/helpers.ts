import {
  AlignMap,
  JustifyMap,
  AlignWithMap,
  JustifyWithMap,
  AlignItems,
  JustifyContent,
} from "./types";

export function mapAlignToProperty(property: AlignWithMap): AlignItems {
  if (property in AlignMap) {
    return AlignMap[property as keyof typeof AlignMap];
  }
  return property as AlignItems;
}

export function mapJustifyToProperty(property: JustifyWithMap): JustifyContent {
  if (property in JustifyMap) {
    return JustifyMap[property as keyof typeof JustifyMap];
  }
  return property as JustifyContent;
}
