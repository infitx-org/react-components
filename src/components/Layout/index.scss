@use "./Layout.scss";
@use "./components/Navbar/Navbar.scss";
