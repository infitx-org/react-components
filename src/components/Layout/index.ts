export { default, SharedProps as LayoutProps } from "./Layout";
