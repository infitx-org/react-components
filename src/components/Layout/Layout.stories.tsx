/* eslint no-console: "off" */
import Layout from "./Layout";

export default {
  title: "Components/Layout",
  component: Layout,
  subcomponents: {
    "Layout.Navbar": Layout.Navbar,
    "Layout.Page": Layout.Page,
    "Layout.SideMenu": Layout.SideMenu,
    "Layout.Content": Layout.Content,
  },
};

const userIcon = (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="40"
    height="40"
    viewBox="0 0 40 40"
  >
    <g>
      <circle cx="20" cy="20" r="20" fill="#fff" />
    </g>
  </svg>
);

const Template = (args) => <Layout {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: (
    <>
      <Layout.Navbar
        title="Layout"
        username="test"
        onUsernameClick={console.log}
        userIcon={userIcon}
      />
      <Layout.Content>
        <Layout.SideMenu>Menu</Layout.SideMenu>
        <Layout.Page>Page</Layout.Page>
      </Layout.Content>
    </>
  ),
};

export const ComposingNavbar = () => (
  <Layout.Navbar title="Composing Navbar">
    <Layout.Navbar.User username="User clickable name">
      <Layout.Navbar.User.Item onClick={console.log}>
        Logout
      </Layout.Navbar.User.Item>
      <Layout.Navbar.User.Item onClick={console.log}>
        Profile
      </Layout.Navbar.User.Item>
    </Layout.Navbar.User>
  </Layout.Navbar>
);
