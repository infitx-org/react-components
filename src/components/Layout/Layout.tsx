import classnames from "classnames";
import Icon from "components/Icon";
import Tooltip from "components/Tooltip";
import TestIcon from "resources/icons/test.svg";
import "./Layout.scss";

export interface SharedProps {
  children: React.ReactNode;
  className?: string;
}

export type LayoutContentProps = SharedProps;

function LayoutContent({ children, className }: LayoutContentProps) {
  return (
    <div className={classnames(["rc-layout__content", className])}>
      {children}
    </div>
  );
}

export type LayoutPageProps = SharedProps;

function LayoutPage({ children, className }: LayoutPageProps) {
  return (
    <div className={classnames(["rc-layout__page", className])}>{children}</div>
  );
}

export type LayoutSideMenuProps = SharedProps;

function LayoutSideMenu({ children, className }: LayoutSideMenuProps) {
  return (
    <div className={classnames(["rc-layout__side-menu", className])}>
      {children}
    </div>
  );
}

export interface LayoutNavbarProps {
  title: string;
  username?: string;
  className?: string;
  onLogoutClick?: () => void;
}

function LayoutNavbar({
  title,
  username = "-",
  className,
  onLogoutClick,
}: LayoutNavbarProps) {
  return (
    <div className={classnames(["rc-layout__navbar", className])}>
      <div className="rc-layout__navbar__controls">
        <a className="rc-layout__navbar__link" href="/">
          {title}
        </a>
      </div>
      <div className="rc-layout__navbar__user">
        <div className="rc-layout__navbar__user__icon">
          <Icon icon={<TestIcon />} fill="#fff" />
        </div>
        <div
          className="rc-layout__navbar__user__name"
          onClick={onLogoutClick}
          role="presentation"
        >
          <Tooltip label="Logout">
            <span>{username}</span>
          </Tooltip>
        </div>
      </div>
    </div>
  );
}

export type LayoutProps = SharedProps;

function Layout({ children, className }: LayoutProps) {
  return <div className={classnames(["rc-layout", className])}>{children}</div>;
}

Layout.Navbar = LayoutNavbar;
Layout.Page = LayoutPage;
Layout.SideMenu = LayoutSideMenu;
Layout.Content = LayoutContent;

export default Layout;
