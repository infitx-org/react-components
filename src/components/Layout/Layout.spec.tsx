import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Layout from "./Layout";

describe("tests the Layout as a whole", () => {
  it("renders the layout components", () => {
    const { container } = render(
      <Layout>
        <Layout.Navbar title="Layout" username="user" />
        <Layout.Content>
          <Layout.SideMenu>Menu</Layout.SideMenu>
          <Layout.Page>Page</Layout.Page>
        </Layout.Content>
      </Layout>
    );
    expect(container.querySelector(".rc-layout")).toBeTruthy();
    expect(container.querySelector(".rc-layout__navbar")).toBeTruthy();
    expect(container.querySelector(".rc-layout__content")).toBeTruthy();
    expect(container.querySelector(".rc-layout__side-menu")).toBeTruthy();
    expect(container.querySelector(".rc-layout__page")).toBeTruthy();
  });
});

describe("tests the Layout component", () => {
  it("renders the layout", () => {
    const { container } = render(<Layout>layout</Layout>);
    expect(container.querySelector(".rc-layout")).toBeTruthy();
    expect(container.querySelector(".rc-layout")?.textContent).toBe("layout");
  });

  it("renders the prop classname on the layout", () => {
    const { container } = render(<Layout className="test">content</Layout>);
    expect(container.querySelector(".rc-layout.test")).toBeTruthy();
  });
});

describe("tests the Layout.Content component", () => {
  it("renders the content", () => {
    const { container } = render(<Layout.Content>content</Layout.Content>);
    expect(container.querySelector(".rc-layout__content")).toBeTruthy();
    expect(container.querySelector(".rc-layout__content")?.textContent).toBe(
      "content"
    );
  });

  it("renders the prop classname on the content", () => {
    const { container } = render(
      <Layout.Content className="test">content</Layout.Content>
    );
    expect(container.querySelector(".rc-layout__content.test")).toBeTruthy();
  });
});

describe("tests the Layout.SideMenu component", () => {
  it("renders the sidemenu", () => {
    const { container } = render(<Layout.SideMenu>side menu</Layout.SideMenu>);
    expect(container.querySelector(".rc-layout__side-menu")).toBeTruthy();
    expect(container.querySelector(".rc-layout__side-menu")?.textContent).toBe(
      "side menu"
    );
  });

  it("renders the prop classname on the sidemenu", () => {
    const { container } = render(
      <Layout.SideMenu className="test">content</Layout.SideMenu>
    );
    expect(container.querySelector(".rc-layout__side-menu.test")).toBeTruthy();
  });
});

describe("tests the Layout.Page component", () => {
  it("renders the page", () => {
    const { container } = render(<Layout.Page>page</Layout.Page>);
    expect(container.querySelector(".rc-layout__page")).toBeTruthy();
    expect(container.querySelector(".rc-layout__page")?.textContent).toBe(
      "page"
    );
  });

  it("renders the prop classname on the page", () => {
    const { container } = render(
      <Layout.Page className="test">content</Layout.Page>
    );
    expect(container.querySelector(".rc-layout__page.test")).toBeTruthy();
  });
});

describe("tests the Layout.Navbar component", () => {
  it("renders the navbar", () => {
    const { container } = render(
      <Layout.Navbar title="Navbar" username="user" />
    );
    expect(container.querySelector(".rc-layout__navbar")).toBeTruthy();
  });

  it("renders the title on the navbar", () => {
    const { container } = render(
      <Layout.Navbar title="Navbar" username="user" />
    );
    expect(
      container.querySelector(".rc-layout__navbar__link")?.textContent
    ).toBe("Navbar");
  });

  it("renders the username on the navbar", () => {
    const { container } = render(
      <Layout.Navbar title="Navbar" username="User" />
    );
    expect(
      container.querySelector(".rc-layout__navbar__block")?.textContent
    ).toBe("User");
  });

  it("renders the prop classname on the navbar", () => {
    const { container } = render(
      <Layout.Navbar title="Layout" username="user" className="test" />
    );
    expect(container.querySelector(".rc-layout__navbar.test")).toBeTruthy();
  });
});

it("renders the layout correctly when multiple props are set", () => {
  const { container } = render(
    <Layout>
      <Layout.Navbar title="Layout" username="user" />
      <Layout.Content>
        <Layout.SideMenu>Menu</Layout.SideMenu>
        <Layout.Page>Page</Layout.Page>
      </Layout.Content>
    </Layout>
  );
  expect(container).toMatchSnapshot();
});
