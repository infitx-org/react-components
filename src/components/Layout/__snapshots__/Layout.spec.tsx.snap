// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the layout correctly when multiple props are set 1`] = `
<div>
  <div
    class="rc-layout"
  >
    <div
      class="rc-layout__navbar"
    >
      <div
        class="rc-layout__navbar__controls"
      >
        <a
          class="rc-layout__navbar__link"
          href="/"
        >
          Layout
        </a>
      </div>
      <div
        class="rc-layout__navbar__user"
      >
        <div
          class="rc-layout__navbar__user__icon"
        >
          <svg
            fill="#fff"
            height="20"
            stroke="transparent"
            width="20"
          />
        </div>
        <div
          class="rc-layout__navbar__user__name"
          role="presentation"
        >
          <span>
            user
          </span>
        </div>
      </div>
    </div>
    <div
      class="rc-layout__content"
    >
      <div
        class="rc-layout__side-menu"
      >
        Menu
      </div>
      <div
        class="rc-layout__page"
      >
        Page
      </div>
    </div>
  </div>
</div>
`;
